# Walmart Sales Analysis — Lesson 1: Data Understanding

## Objective

The goal of this lesson is to understand the structure and quality of the
Walmart datasets before performing any data cleaning or analysis.

### We will investigate:

- Dataset size and structure
- Column names and data types
- Sample records
- Store distribution
- Holiday distribution
- Missing values
- Duplicate records
- Relationships between the datasets

In [113]:
# Import libraries required for data analysis
import pandas as pd
import numpy as np

## 1. Load the Datasets

### Question

What datasets will we use for the Walmart sales analysis?

In [116]:
# Load the Features, Sales, and Stores datasets
features = pd.read_csv("/content/drive/MyDrive/Features data set.csv")
sales = pd.read_csv("/content/drive/MyDrive/sales data-set.csv")
stores = pd.read_csv("/content/drive/MyDrive/stores data-set.csv")

## 2. Understand Dataset Dimensions

### Question

How many rows and columns are present in each dataset?

Understanding the dimensions helps us know the size of the data we will be working with.

In [117]:
# Check the number of rows and columns in each dataset
print("Features:", features.shape)
print("Sales:", sales.shape)
print("Stores:", stores.shape)

Features: (8190, 12)
Sales: (421570, 5)
Stores: (45, 3)


### Interpretation

- The Features dataset contains 8,190 rows and 12 columns.
- The Sales dataset contains 421,570 rows and 5 columns.
- The Stores dataset contains 45 rows and 3 columns.

The Sales dataset contains the largest number of records, while the Stores dataset is much smaller because each store is represented by a single record.

## 3. Understand the Dataset Columns

### Question

What columns are available in each dataset?

Knowing the column names helps us understand what information is available before beginning the analysis.

In [118]:
# Display the column names of each dataset
print("Features columns:")
print(features.columns)

print("\nSales columns:")
print(sales.columns)

print("\nStores columns:")
print(stores.columns)

Features columns:
Index(['Store', 'Date', 'Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2',
       'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment',
       'IsHoliday'],
      dtype='object')

Sales columns:
Index(['Store', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday'], dtype='object')

Stores columns:
Index(['Store', 'Type', 'Size'], dtype='object')


### Interpretation

The three datasets contain different types of information:

- **Features:** Contains store-level information such as temperature, fuel price,
  markdown values, CPI, unemployment, and holiday status.
- **Sales:** Contains weekly sales information for each store and department.
- **Stores:** Contains store-level information including store type and store size.

The `Store` and `Date` columns are shared between the Features and Sales datasets,
while `Store` connects the Stores dataset with the other datasets.

### Key Observation

The datasets are related through the `Store` column, which will allow us to
combine information from different datasets in later analysis.

## 4. Understand Data Types

### Question

What data types are used for the columns in each dataset?

Understanding data types is important because different types of data require
different analysis and cleaning methods.

In [119]:
# Display the data type of each column in all three datasets
print("Features data types:")
print(features.dtypes)

print("\nSales data types:")
print(sales.dtypes)

print("\nStores data types:")
print(stores.dtypes)

Features data types:
Store             int64
Date             object
Temperature     float64
Fuel_Price      float64
MarkDown1       float64
MarkDown2       float64
MarkDown3       float64
MarkDown4       float64
MarkDown5       float64
CPI             float64
Unemployment    float64
IsHoliday          bool
dtype: object

Sales data types:
Store             int64
Dept              int64
Date             object
Weekly_Sales    float64
IsHoliday          bool
dtype: object

Stores data types:
Store     int64
Type     object
Size      int64
dtype: object


### Interpretation

- `Store`, `Dept`, and `Size` are stored as integers because they contain whole-number values.
- Temperature, Fuel Price, Markdown values, CPI, Unemployment, and Weekly Sales are stored as floating-point numbers because they can contain decimal values.
- `IsHoliday` is stored as a Boolean value (`True` or `False`).
- `Type` and `Date` are currently stored as `object`.

### Key Observation

The `Date` column is stored as `object` rather than a date/time data type.
This may need to be converted to a proper datetime format before performing
time-based analysis.

## 5. Preview the Sales Data

### Question

What does the Sales dataset look like at the beginning and end of the dataset?

Viewing sample records helps us understand the structure and values before performing analysis.

In [120]:
# Display the first 3 rows of the Sales dataset
sales.head(3)

,Store,Dept,Date,Weekly_Sales,IsHoliday
0,1,1,05/02/2010,24924.50,False
1,1,1,12/02/2010,46039.49,True
2,1,1,19/02/2010,41595.55,False


In [121]:
# Display the last 3 rows of the Sales dataset
sales.tail(3)

,Store,Dept,Date,Weekly_Sales,IsHoliday
421567,45,98,12/10/2012,1061.02,False
421568,45,98,19/10/2012,760.01,False
421569,45,98,26/10/2012,1076.80,False


In [122]:
# Display 5 randomly selected rows from the Sales dataset
sales.sample(5)

,Store,Dept,Date,Weekly_Sales,IsHoliday
186242,19,91,11/11/2011,60270.76,False
345954,36,95,11/06/2010,51956.47,False
55989,6,59,31/08/2012,167.03,False
1927,1,14,27/05/2011,12746.66,False
179229,19,16,12/11/2010,6428.03,False


### Interpretation

The first, last, and random records confirm that the Sales dataset contains
weekly sales information for individual stores and departments.

The dataset includes:

- Store
- Department
- Date
- Weekly Sales
- Holiday status

Sampling different parts of the dataset helps confirm that the records follow
a consistent structure.

## 6. Explore Store Types

### Question

How many different store types are present in the Stores dataset?

In [123]:
# Find the unique store types
stores["Type"].unique()

array(['A', 'B', 'C'], dtype=object)

### Interpretation

The Stores dataset contains three store types:

- Type A
- Type B
- Type C

### Question

How many stores belong to each store type?

In [124]:
# Count the number of stores in each store type
stores["Type"].value_counts()

,count
Type,
A,22
B,17
C,6


### Interpretation

The store distribution is:

- Type A: 22 stores
- Type B: 17 stores
- Type C: 6 stores

Type A stores represent the largest group, while Type C stores are the smallest group.

### Question

What percentage of stores belong to each store type?

In [125]:
# Calculate the percentage distribution of store types
stores["Type"].value_counts(normalize=True) * 100

,proportion
Type,
A,48.888889
B,37.777778
C,13.333333


### Interpretation

The percentage distribution shows the relative share of each store type
within the 45 stores in the dataset.

This gives us a clearer view of the store structure than counts alone.

## 7. Explore Holiday Distribution

### Question

How many sales records occurred during holiday weeks compared with non-holiday weeks?

Understanding the holiday distribution helps us understand how much of the sales data is associated with holiday periods.

In [128]:
# Count holiday and non-holiday sales records
sales["IsHoliday"].value_counts()

,count
IsHoliday,
False,391909
True,29661


### Interpretation

- 391,909 sales records are from non-holiday weeks.
- 29,661 sales records are from holiday weeks.

Non-holiday records make up the majority of the Sales dataset.

### Question

What percentage of sales records occurred during holiday and non-holiday weeks?

In [129]:
# Calculate the percentage distribution of holiday and non-holiday records
sales["IsHoliday"].value_counts(normalize=True) * 100

,proportion
IsHoliday,
False,92.964158
True,7.035842


### Interpretation

Approximately 92.96% of sales records are from non-holiday weeks,
while 7.04% are from holiday weeks.

This shows that holiday weeks represent a relatively small portion of the
overall Sales dataset.

## 8. Investigate Missing Values

### Question

Are there any missing values in the Features dataset?

Identifying missing values is an important step before performing data cleaning or analysis.

In [142]:
# Count missing values in each column of the Features dataset
print(features.isnull().sum())

Store              0
Date               0
Temperature        0
Fuel_Price         0
MarkDown1       4158
MarkDown2       5269
MarkDown3       4577
MarkDown4       4726
MarkDown5       4140
CPI              585
Unemployment     585
IsHoliday          0
dtype: int64


### Interpretation

The Features dataset contains missing values in seven columns:

- MarkDown1: 4,158 missing values
- MarkDown2: 5,269 missing values
- MarkDown3: 4,577 missing values
- MarkDown4: 4,726 missing values
- MarkDown5: 4,140 missing values
- CPI: 585 missing values
- Unemployment: 585 missing values

The remaining columns have no missing values.

### Key Observation

The Markdown columns contain substantially more missing values than CPI and
Unemployment. We will investigate the percentage of missing values before
deciding how these columns should be handled.

### Question

What percentage of values are missing in each column of the Features dataset?

In [133]:
# Calculate the percentage of missing values in each Features column
missing_percentage = (features.isnull().sum() / len(features)) * 100

print(missing_percentage)

Store            0.000000
Date             0.000000
Temperature      0.000000
Fuel_Price       0.000000
MarkDown1       50.769231
MarkDown2       64.334554
MarkDown3       55.885226
MarkDown4       57.704518
MarkDown5       50.549451
CPI              7.142857
Unemployment     7.142857
IsHoliday        0.000000
dtype: float64


### Interpretation

The missing-value percentages show a clear difference between the columns:

- `MarkDown1`: 50.77% missing
- `MarkDown2`: 64.33% missing
- `MarkDown3`: 55.89% missing
- `MarkDown4`: 57.70% missing
- `MarkDown5`: 50.55% missing
- `CPI`: 7.14% missing
- `Unemployment`: 7.14% missing

The five Markdown columns have more than 50% missing values, while CPI and
Unemployment have a much lower missing percentage.

### Key Observation

The Markdown columns require further investigation before deciding whether to
remove, fill, or otherwise handle their missing values.

### Question

Which columns contain more than 50% missing values?

Instead of checking each percentage manually, we can use Boolean filtering
to identify these columns automatically.

In [134]:
# Identify columns with more than 50% missing values
missing_percentage[missing_percentage > 50]

,0
MarkDown1,50.769231
MarkDown2,64.334554
MarkDown3,55.885226
MarkDown4,57.704518
MarkDown5,50.549451


### Interpretation

All five Markdown columns contain more than 50% missing values.

- `MarkDown2` has the highest missing percentage at 64.33%.
- `MarkDown5` has the lowest missing percentage among the Markdown columns at 50.55%.

These columns should not be automatically removed or filled without first
understanding what the missing values represent.

### Question

Which columns contain less than 10% missing values?

This helps us distinguish columns with relatively low missingness from columns
with substantial missingness.

In [135]:
# Identify columns with less than 10% missing values
missing_percentage[missing_percentage < 10]

,0
Store,0.000000
Date,0.000000
Temperature,0.000000
Fuel_Price,0.000000
CPI,7.142857
Unemployment,7.142857
IsHoliday,0.000000


### Interpretation

The following columns have less than 10% missing values:

- `Store`: 0% missing
- `Date`: 0% missing
- `Temperature`: 0% missing
- `Fuel_Price`: 0% missing
- `CPI`: 7.14% missing
- `Unemployment`: 7.14% missing
- `IsHoliday`: 0% missing

This shows that most non-Markdown columns are complete, while CPI and
Unemployment have a relatively small amount of missing data.

## 9. Check for Duplicate Records

### Question

Are there any completely duplicated rows in the Features, Sales, or Stores datasets?

Duplicate records can lead to incorrect counts, totals, and analysis results.

In [136]:
# Count completely duplicated rows in each dataset
print("Features duplicates:", features.duplicated().sum())
print("Sales duplicates:", sales.duplicated().sum())
print("Stores duplicates:", stores.duplicated().sum())

Features duplicates: 0
Sales duplicates: 0
Stores duplicates: 0


### Interpretation

No completely duplicated rows were found in any of the three datasets:

- Features: 0 duplicates
- Sales: 0 duplicates
- Stores: 0 duplicates

This indicates that there are no exact duplicate records based on all columns
within each dataset.

### Question

Does the Sales dataset contain duplicate Store + Department + Date combinations?

This checks whether the same store, department, and date appear more than once.

In [137]:
# Check for duplicate Store + Department + Date combinations
sales.duplicated(
    subset=["Store", "Dept", "Date"]
).sum()

np.int64(0)

### Interpretation

There are no duplicate combinations of `Store`, `Dept`, and `Date` in the
Sales dataset.

This means each Store + Department + Date combination appears only once,
which supports the expected structure of the Sales data.

## 10. Validate Relationships Between Datasets

### Question

Does every Store ID in the Sales dataset exist in the Stores dataset?

This check helps verify that the Store column can be reliably used to connect
the Sales and Stores datasets.

In [138]:
# Check whether every Store ID in Sales exists in Stores
sales["Store"].isin(stores["Store"]).all()

np.True_

### Interpretation

Every Store ID in the Sales dataset is present in the Stores dataset.

This confirms that the `Store` column can be used as a valid connection
between the Sales and Stores datasets.

### Question

Does every Store ID in the Stores dataset appear in the Sales dataset?

Checking both directions helps confirm that the Store IDs are consistently
represented across the two datasets.

In [139]:
# Check whether every Store ID in Stores exists in Sales
stores["Store"].isin(sales["Store"]).all()

np.True_

### Interpretation

Every Store ID in the Stores dataset is also present in the Sales dataset.

Together with the previous check, this confirms that:

- Every Store in Sales exists in Stores.
- Every Store in Stores appears in Sales.

Therefore, the `Store` column provides a consistent connection between the
Sales and Stores datasets.

# Lesson 1 — Key Findings

## Dataset Overview

- The Features dataset contains 8,190 rows and 12 columns.
- The Sales dataset contains 421,570 rows and 5 columns.
- The Stores dataset contains 45 rows and 3 columns.

## Data Structure

- Features contains store, date, weather, economic, Markdown, and holiday information.
- Sales contains weekly sales by store and department.
- Stores contains store type and store size information.
- The `Store` column connects the datasets.

## Store Distribution

- Type A: 22 stores
- Type B: 17 stores
- Type C: 6 stores

## Holiday Distribution

- Non-holiday records: approximately 92.96%
- Holiday records: approximately 7.04%

## Missing Values

- The five Markdown columns contain more than 50% missing values.
- `MarkDown2` has the highest missing percentage at 64.33%.
- `CPI` and `Unemployment` each have 7.14% missing values.
- Store, Date, Temperature, Fuel_Price, and IsHoliday contain no missing values.

## Duplicate Checks

- No completely duplicated rows were found in Features, Sales, or Stores.
- No duplicate Store + Department + Date combinations were found in Sales.

## Relationship Validation

- Every Store ID in Sales exists in Stores.
- Every Store ID in Stores exists in Sales.

## Overall Conclusion

The datasets have a consistent structure and can be connected through the
`Store` column. The main data-quality issue identified during this initial
investigation is the high level of missing values in the Markdown columns.

No cleaning decisions are made at this stage. The identified issues will be
investigated further before data cleaning and analysis.